File: src/types/health.ts
// HoloSelf OS — Health Domain Types

export interface SupplementEntry {
  id?: number;
  name: string;
  dosage: string;
  taken_at: string; // ISO 8601
  category: "morning" | "night" | "as_needed";
  notes?: string;
}

export interface VitalEntry {
  id?: number;
  vital_type: "heart_rate" | "hrv" | "sleep_score" | "stress_level" | "wpm";
  value: number;
  unit: string;
  recorded_at: string; // ISO 8601
  source: "manual" | "wearable" | "webcam";
}

export interface ClinicalResult {
  marker: string;
  value: number;
  unit: string;
  reference_range: string;
  status: "normal" | "low" | "high" | "critical";
}

export interface OcrResult {
  patient_name?: string;
  date?: string;
  lab?: string;
  markers: ClinicalResult[];
  raw_text?: string;
}

export interface AgentMessage {
  text: string;
  category: "supplement_reminder" | "health_insight" | "calm_nudge" | "schedule";
  priority: "low" | "medium" | "high";
  action: AgentAction | null;
}

export interface AgentAction {
  action_type: "log_supplement" | "schedule_exam" | "open_timer";
  payload: Record<string, unknown>;
}

export interface HealthTimelineEntry {
  timestamp: string;
  event_type: string;
  label: string;
  value?: number;
}

export interface HealthScheduleEntry {
  id?: number;
  exam_type: string;
  reason: string;
  scheduled_date: string;
  triggered_by?: string;
  completed: boolean;
}

export interface SystemStatus {
  version: string;
  db_connected: boolean;
  gemini_configured: boolean;
  voice_available: boolean;
  timezone: string;
  uptime_seconds: number;
}

File: src-tauri/src/commands/health.rs
use serde::{Deserialize, Serialize};
use tauri::State;
use crate::db::DbState;

#[derive(Debug, Serialize, Deserialize)]
pub struct SupplementEntry {
    pub id: Option<i64>,
    pub name: String,
    pub dosage: String,
    pub taken_at: String,   // ISO 8601 datetime
    pub category: String,   // morning | night | as_needed
    pub notes: Option<String>,
}

#[derive(Debug, Serialize, Deserialize)]
pub struct VitalEntry {
    pub id: Option<i64>,
    pub vital_type: String, // heart_rate | hrv | sleep_score | stress_level | wpm
    pub value: f64,
    pub unit: String,
    pub recorded_at: String,
    pub source: String,     // manual | wearable | webcam
}

#[derive(Debug, Serialize, Deserialize)]
pub struct HealthTimelineEntry {
    pub timestamp: String,
    pub event_type: String,
    pub label: String,
    pub value: Option<f64>,
}

/// Log a supplement intake
#[tauri::command]
pub async fn log_supplement(
    state: State<'_, DbState>,
    entry: SupplementEntry,
) -> Result<i64, String> {
    let db = state.0.lock().map_err(|e| e.to_string())?;
    db.insert_supplement(&entry).map_err(|e| e.to_string())
}

/// Get supplement log for a date range
#[tauri::command]
pub async fn get_supplement_log(
    state: State<'_, DbState>,
    from: String,
    to: String,
) -> Result<Vec<SupplementEntry>, String> {
    let db = state.0.lock().map_err(|e| e.to_string())?;
    db.get_supplements(&from, &to).map_err(|e| e.to_string())
}

/// Get unified health timeline
#[tauri::command]
pub async fn get_health_timeline(
    state: State<'_, DbState>,
    from: String,
    to: String,
) -> Result<Vec<HealthTimelineEntry>, String> {
    let db = state.0.lock().map_err(|e| e.to_string())?;
    db.get_health_timeline(&from, &to).map_err(|e| e.to_string())
}

/// Log a vital sign measurement
#[tauri::command]
pub async fn log_vital(
    state: State<'_, DbState>,
    entry: VitalEntry,
) -> Result<i64, String> {
    let db = state.0.lock().map_err(|e| e.to_string())?;
    db.insert_vital(&entry).map_err(|e| e.to_string())
}
